L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,504 @@
+
+  // Propagates an error encountered in the provider layer to the internal TLS stack.
+  //
+  // @param cert_name The watching certificate name that the caller
+  // wants to notify about when encountering error.
+  // @param root_cert_error The error that the caller encounters when reloading
+  // root certificates.
+  // @param identity_cert_error The error that the caller encounters when
+  // reloading identity certificates.
+  void SetErrorForCert(const std::string& cert_name,
+                       absl::Status root_cert_error,
+                       absl::Status identity_cert_error);
+
+ private:
+  std::unique_ptr<TlsCertificateDistributor> distributor_;
+};
+
+/** ------------------------------------ Provider ------------------------------------ */
+/** Provides identity credentials and root certificates.
+*/
+// This represents a potential decoupling of roots and identity chains, with
+// further extension points for something like SPIFFE bundles
+class CertificateProviderInterface {
+ public:
+  virtual ~CertificateProviderInterface() = default;
+  // Provider implementations MUST provide a LoadAndSetCredentialsCallback that
+  // will be called by the internal stack. This will be invoked when a new
+  // certificate name is starting to be used internally, or when a certificate
+  // name is no longer being used internally. When being invoked for a new
+  // certificate, this callback should call SetRootCertificates or
+  // SetIdentityChainAndPrivateKey to do the initial population the certificate
+  // data in the internal stack.
+  // 
+  // For the parameters in the callback function: cert_name The name of the
+  // certificates being watched.  type The type of certificates being watched.
+  virtual void LoadAndSetCredentialsCallback(std::string name, CredentialType type) = 0;
+
+  // Must be called after the constructor.
+  // Does important internal setup steps.
+  virtual void Init() final;
+
+  enum CredentialType {
+    RootCertificates,
+    IdentityChainAndPrivateKey
+  }
+
+protected:
+  // Sets the root certificates based on their name.
+  virtual void SetRootCertificates(const std::string& name, std::string pem_root_certificates) final;
+
+  // Sets the identity chain and private key based on their name.
+  virtual void SetIdentityChainAndPrivateKey(const std::string& name, grpc_core::PemKeyCertPairList pem_key_cert_pairs) final;
+
+  // Propagates an error encountered in the provider layer to the internal TLS stack.
+  virtual void SetError(const std::string& name, CredentialType type, absl::Status error) final;
+
+ private:
+  std::unique_ptr<TlsCertificateDistributor> distributor_;
+};
+
+// A basic CertificateProviderInterface implementation that will load credential
+// data from static string during initialization. This provider will always
+// return the same certificate data for all cert names, and reloading is not supported.
+class StaticDataCertificateProvider
```

Comment:
From review, consider also an in-memory provider that lets an application totally own the different buffers.

Reply:
Also had argument that we should just make it easy to implement new cert providers, which seemed to be what we'd like to go with.